File: chaincode/src/main/java/edu/tsp/ProductProperties.java
/*
 * SPDX-License-Identifier: Apache-2.0
 */

package edu.tsp;

import com.owlike.genson.annotation.JsonProperty;
import org.hyperledger.fabric.contract.annotation.DataType;
import org.hyperledger.fabric.contract.annotation.Property;

import java.util.List;
import java.util.Objects;

/**
 * ProductProperties structure used for handling result of query
 *
 */
@DataType()
public final class ProductProperties {
    @Property()
    private final String productId;

    @Property()
    private final List<ProductProperty> properties;

    public ProductProperties(@JsonProperty("productId") final String productId,
                             @JsonProperty("properties") final List<ProductProperty> properties) {
        this.productId = productId;
        this.properties = properties;
    }

    public String getProductId() {
        return productId;
    }

    public List<ProductProperty> getProperties() {
        return properties;
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) {
            return true;
        }

        if ((obj == null) || (getClass() != obj.getClass())) {
            return false;
        }

        ProductProperties other = (ProductProperties) obj;

        return productId.equals(other.productId) && Objects.deepEquals(properties, other.properties);
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.getProductId(), this.getProperties());
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append(this.getClass().getSimpleName());
        builder.append("@");
        builder.append(Integer.toHexString(hashCode()));
        builder.append(" [productId=").append(productId);
        for (ProductProperty property : properties) {
            builder.append(", ")
                    .append(property.getName())
                    .append("=")
                    .append(property.getValue());
        }
        builder.append("]");

        return builder.toString();
    }
}

File: chaincode/tsp/javascript/property/utils/ProductProperties.js
/*
 * SPDX-License-Identifier: Apache-2.0
 */



/**
 * ProductProperties structure used for handling result of query
 */
class ProductProperties {
    /**
     * @param {string} productId
     * @param {Array<object>} properties - An array of ProductProperty objects,
     * where each object has 'name' and 'value' properties.
     */
    constructor(productId, properties) {
        this.productId = productId;
        this.properties = properties;
    }

    /**
     * Gets the product ID.
     * @returns {string} The product ID.
     */
    getProductId() {
        return this.productId;
    }

    /**
     * Gets the list of product properties.
     * @returns {Array<object>} The list of product properties.
     */
    getProperties() {
        return this.properties;
    }

    /**
     * Checks if this ProductProperties object is equal to another object.
     * @param {object} obj The object to compare with.
     * @returns {boolean} True if the objects are equal, false otherwise.
     */
    equals(obj) {
        if (this === obj) {
            return true;
        }

        if (!obj || obj.constructor !== this.constructor) {
            return false;
        }

        const other = obj;

        // Deep comparison for the 'properties' array
        const propertiesEqual = (Array.isArray(this.properties) && Array.isArray(other.properties)) &&
                                this.properties.length === other.properties.length &&
                                this.properties.every((prop, index) => {
                                    const otherProp = other.properties[index];
                                    return prop.name === otherProp.name && prop.value === otherProp.value;
                                });

        return this.productId === other.productId && propertiesEqual;
    }

    /**
     * Generates a hash code for this object.
     * (Note: JavaScript doesn't have a direct equivalent of Java's hashCode for object identity.
     * This is a simple hash based on property values.)
     * @returns {string} A string representation of the hash.
     */
    hashCode() {
        // A simple way to create a "hash" in JS, good for logging/debugging
        return JSON.stringify({ productId: this.productId, properties: this.properties }).split('').reduce((a, b) => {
            a = ((a << 5) - a) + b.charCodeAt(0);
            return a & a;
        }, 0).toString();
    }


    /**
     * Returns a string representation of the object.
     * @returns {string} The string representation.
     */
    toString() {
        let builder = `${this.constructor.name}@${this.hashCode()} [productId=${this.productId}`;
        for (const property of this.properties) {
            builder += `, ${property.name}=${property.value}`;
        }
        builder += `]`;
        return builder;
    }
}

module.exports = ProductProperties;